### PR TITLE
Framework in place for salt-ssh to windows

### PR DESCRIPTION
**Pull request by thatch45**
This has proven to be a monster, we got most of it working but it just keeps cluttering up the salt-ssh code. I want this to make it into Nitrogen so that we have the hooks into an external library we are making to make things go smoothly, but there is still a lot of work left until this is ready.
Hopefully we can get this working fully soonish!

### CODE REVIEW

File: salt/client/ssh/__init__.py [hunk 1]
```diff
@@ -1040,35 +1053,40 @@ def _cmd_str(self):
...
-        if not self.tty:
+        if not self.tty and not self.winrm:
             return self.shell.exec_cmd(cmd_str)
 
         # Write the shim to a temporary file in the default temp directory
-        with tempfile.NamedTemporaryFile(mode='w',
+        with tempfile.NamedTemporaryFile(mode='w+b',
                                          prefix='shim_',
                                          delete=False) as shim_tmp_file:
             shim_tmp_file.write(cmd_str)
```
Comment (s0undt3ch) [inline salt/client/ssh/__init__.py]: This will not be py3 compat but just get it in and we'll fix it!

File: salt/client/ssh/shell.py [hunk 2]
```diff
@@ -46,6 +46,22 @@ def gen_key(path):
...
+    '''
+    Return the correct shell interface for the target system
+    '''
+    if kwargs['winrm']:
+        try:
+            import saltwinshell
+            shell = saltwinshell.Shell(opts, **kwargs)
+        except ImportError:
+            log.error('The saltwinshell library is not available')
+            sys.exit(1)
```
Comment (s0undt3ch) [inline salt/client/ssh/shell.py]: Use salt exitcodes module?

**Comment by cachedout** [comment]
I'm going to see if @thatch45 wants to fix those really quickly this morning. Otherwise, I'll get this in.

**Comment by thatch45** [comment]
Thank you gentlemen, fixes submitted!

**Pull request merged**